server/src/main/java/org/elasticsearch/common/util/concurrent/ListenableFuture.java

```diff
@@ -30,31 +30,33 @@
 public final class ListenableFuture<V> extends BaseFuture<V> implements ActionListener<V> {
 
     private volatile boolean done = false;
-    private final List<Tuple<ActionListener<V>, ExecutorService>> listeners = new ArrayList<>();
-
+    private List<Tuple<ActionListener<V>, ExecutorService>> listeners;
 
     /**
      * Adds a listener to this future. If the future has not yet completed, the listener will be
-     * notified of a response or exception in a runnable submitted to the ExecutorService provided.
+     * notified of a response or exception on the thread completing this future.
      * If the future has completed, the listener will be notified immediately without forking to
      * a different thread.
      */
-    public void addListener(ActionListener<V> listener, ExecutorService executor) {
-        addListener(listener, executor, null);
+    public void addListener(ActionListener<V> listener) {
+        addListener(listener, null, null);
     }
 
     /**
      * Adds a listener to this future. If the future has not yet completed, the listener will be
-     * notified of a response or exception in a runnable submitted to the ExecutorService provided.
+     * notified of a response or exception in a runnable submitted to the ExecutorService provided
+     * if one is provided. If a null executor is provided the listener will be executed directly
+     * on the thread completing the future.
      * If the future has completed, the listener will be notified immediately without forking to
      * a different thread.
      *
      * It will apply the provided ThreadContext (if not null) when executing the listening.
      */
-    public void addListener(ActionListener<V> listener, ExecutorService executor, ThreadContext threadContext) {
+    public void addListener(ActionListener<V> listener, @Nullable ExecutorService executor, ThreadContext threadContext) {
+        assert executor != EsExecutors.newDirectExecutorService() : "using direct executor here instead of null is needless overhead";
```

Comment:
It would be overhead with the code as it stands now, unless we add a check that compares the executor to the constant.

API wise, to me a `null` check seemed more obvious than comparing with that constant (also I suppose it's technically faster too not that that matters too much here). Someone just ramping up on the code certainly would have an easier time understanding to pass `null` for not forking that to look up that constant (which isn't even public ) and currently only returned by `org.elasticsearch.common.util.concurrent.EsExecutors#newDirectExecutorService` (the name on this thing is unfortunate since it suggests instantiating the direct executor service and should probably be changed also)?

Reply:
Come to think of it, it would be weird to pass the direct executor here and then not actually use it when thinking about the reasoning in https://github.com/elastic/elasticsearch/pull/72087#discussion_r620061385 ?

Reply:
Oh wait I missed this thread, ignore my previous top-level comment

Reply:
Yes the name `newDirectExecutorService` is deeply confusing, we should fix that, let's just expose the constant to make it clear it's the same instance every time.

I disagree that `null` is easier to grok. Personally if I see a method requiring an `EsExecutorService` I expect to be able to pass in a direct executor service if I don't want to fork; `@Nullable` annotations are pretty opaque and fragile in comparison.

Reply:
... and the decision whether to use the direct service or not is definitely a function of whether it's equivalent to just running the thing directly or not. Here we know it is today. If we needed it not to be in future then we could search for all the places we have special-case behaviour for `EsExecutors#DIRECT_EXECUTOR_SERVICE`, but we might well miss places like this if we used `null`. (acking that we at least mention it in an assertion so we probably wouldn't miss it, but still...)

Reply:
> I disagree that null is easier to grok. 

I mean you need to know about the direct executor service (could obvious link in the Javadocs) to be able to run without forking while keeping the thread context around, while allowing to use `null` for this makes it very obvious.

> If we needed it not to be in future then we could search for all the places we have special-case behaviour for EsExecutors#DIRECT_EXECUTOR_SERVICE, but we might well miss places like this if we used null

Not sure I understand that argument. If we just use `null` and not allow the direct executor here then the problem of special casing for it doesn't even come up to begin with.

To me this is mostly a stylistic thing. I just like `null` because it's the least code/docs and executes the fastest (albeit to a degree that isn't relevant here) which I find aesthetically pleasing :)
But if you insist, happy to move to the constant instead and compare with it.

Reply:
Thanks, yes, I'd much prefer that in the context of this PR. Feel free to open another one to have a broader discussion about it tho.

Reply:
:white_flag: https://github.com/elastic/elasticsearch/pull/72087/commits/64882b2a938a75068278b588f3fd74f8e944a1cd 
Even though I still like `null` better this was well worth it for the cleanup elsewhere :)